### PR TITLE
Improve state lock tracking and report on error

### PR DESCRIPTION
**Pull request by jiacheliu3**
### What changes are proposed in this pull request?

1. Change tracking threads to keeping thread read lock holder count
2. Log jstack on state lock error so we can further check who are holding the state lock

### Why are the changes needed?

Make debugging state lock issues slightly easier.

### Does this PR introduce any user facing changes?

A chunky jstack output will appear in the master.log, which may look intimidating at first glance. This is good because I want people to notice and send the output to us.

### CODE REVIEW

File: core/server/common/src/main/java/alluxio/master/StateLockManager.java [hunk 1]
```diff
@@ -158,12 +167,36 @@ public LockResource lockShared() throws InterruptedException {
     }
     // Register thread for interrupt cycle.
     mSharedWaitersAndHolders.add(Thread.currentThread());
-    // Grab the lock interruptibly.
-    mStateLock.readLock().lockInterruptibly();
+    String threadName = Thread.currentThread().getName();
+    mSharedLockHolders.computeIfAbsent(threadName, k -> new LongAdder()).increment();
+    if (mSharedLockHolders.get(threadName).longValue() > mLogThreshold) {
+      Exception e = new Exception("Thread recursion is deeper than " + mLogThreshold);
```
Comment (jiacheliu3) [inline core/server/common/src/main/java/alluxio/master/StateLockManager.java:173]: this ex is just for logging

File: core/server/common/src/main/java/alluxio/master/StateLockManager.java [hunk 2]
```diff
@@ -158,12 +167,36 @@ public LockResource lockShared() throws InterruptedException {
...
+      LOG.warn("Current thread is {}. All state lock holders are {}",
+          threadName, mSharedLockHolders, e);
+    }
+    try {
+      // Grab the lock interruptibly.
+      mStateLock.readLock().lockInterruptibly();
+    } catch (Error e) {
+      // An Error is thrown when the lock is acquired 65536 times, log the jstack before exiting
+      LOG.error("Logging all thread stacks before exiting", e);
+      ThreadUtils.logAllThreads();
```
Comment (jiacheliu3) [inline core/server/common/src/main/java/alluxio/master/StateLockManager.java:183]: it's almost impossible for the support to get in and take jstack when this issue happens, so i'm just dumping before exit. In fact I figure this should be the best approach for many similar situations. FYI @yuzhu

File: core/server/common/src/main/java/alluxio/master/StateLockManager.java [hunk 3]
```diff
@@ -158,12 +167,36 @@ public LockResource lockShared() throws InterruptedException {
...
     // Return the resource.
     // Register an action to remove the thread from holders registry before releasing the lock.
     return new LockResource(mStateLock.readLock(), false, false, () -> {
-      mSharedWaitersAndHolders.remove(Thread.currentThread());
+      // this is invoked in the same thread at the end of try-with-resource so should be okay
+      Thread removedFrom = Thread.currentThread();
+      mSharedLockHolders.computeIfPresent(removedFrom.getName(), (k, v) -> {
+        mSharedWaitersAndHolders.remove(Thread.currentThread());
+        if (v.longValue() <= 1L) {
+          return null;
```
Comment (jiacheliu3) [inline core/server/common/src/main/java/alluxio/master/StateLockManager.java:194]: remove the thread from book-keeping so the map doesn't grow too big

File: core/server/common/src/main/java/alluxio/master/StateLockManager.java [hunk 4]
```diff
@@ -143,7 +154,7 @@ public LockResource lockShared() throws InterruptedException {
       final int readLockCount = mStateLock.getReadLockCount();
       if (readLockCount > READ_LOCK_COUNT_HIGH) {
         SAMPLING_LOG.info("Read Lock Count Too High: {} {}", readLockCount,
-            mSharedWaitersAndHolders);
+            mSharedLockHolders);
```
Comment (elega) [inline core/server/common/src/main/java/alluxio/master/StateLockManager.java:157]: can this be printed properly?

File: core/server/common/src/main/java/alluxio/master/StateLockManager.java (hunk 2 above)
Comment (elega) [reply core/server/common/src/main/java/alluxio/master/StateLockManager.java:185]: should we do this for this "65536" issue alone?

File: core/server/common/src/main/java/alluxio/master/StateLockManager.java (hunk 4 above)
Comment (jiacheliu3) [reply core/server/common/src/main/java/alluxio/master/StateLockManager.java:157]: yes, in the form of `{key=value}`

File: core/server/common/src/main/java/alluxio/master/StateLockManager.java (hunk 2 above)
Comment (jiacheliu3) [reply core/server/common/src/main/java/alluxio/master/StateLockManager.java:185]: Yes the 65536 issue throws `Error`
```
2023-02-07 01:39:33,379 ERROR ProcessUtils - Uncaught exception while running Alluxio master @sj1hdp485.corp.adobe.com:19998, stopping it and exiting. Exception "java.lang.Error: Maximum lock count exceeded", Root Cause "java.lang.Error: Maximum lock count exceeded"
java.lang.Error: Maximum lock count exceeded
	at java.base/java.util.concurrent.locks.ReentrantReadWriteLock$Sync.fullTryAcquireShared(ReentrantReadWriteLock.java:535)
	at java.base/java.util.concurrent.locks.ReentrantReadWriteLock$Sync.tryAcquireShared(ReentrantReadWriteLock.java:494)
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1344)
	at java.base/java.util.concurrent.locks.ReentrantReadWriteLock$ReadLock.lockInterruptibly(ReentrantReadWriteLock.java:783)
	at alluxio.master.StateLockManager.lockShared(StateLockManager.java:162)
	at alluxio.master.AbstractMaster.createJournalContext(AbstractMaster.java:141)
	at ...
```

**Review by elega — APPROVED** [review]

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Comment by jiacheliu3** [comment]
alluxio-bot, cherry-pick this to branch-2.9 please

**Comment by jiacheliu3** [comment]
alluxio-bot, cherry-pick this to branch-2.8 please

File: core/server/common/src/test/java/alluxio/master/StateLockManagerTest.java [hunk 5]
```diff
@@ -140,6 +139,7 @@ public void testExclusiveOnlyMode() throws Throwable {
   }
 
   @Test
+  // TODO(jiacheng): run this test before committing
```
Comment (Xenorith) [inline core/server/common/src/test/java/alluxio/master/StateLockManagerTest.java:142]: we should put in a static check for no TODO comments 😭

Comment (jiacheliu3) [reply core/server/common/src/test/java/alluxio/master/StateLockManagerTest.java:142]: oh shoot i forgot to remove this TODO... These UTs pass before I merged so we are okay per se in terms of quality.

**Pull request merged**